java/flight/src/test/java/org/apache/arrow/flight/TestLeak.java

```diff
@@ -0,0 +1,183 @@
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.Float8Vector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.types.FloatingPointPrecision;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+import org.junit.Test;
+
+/**
+ * Tests for scenarios where Flight could leak memory.
+ */
+public class TestLeak {
+
+  private static final int ROWS = 2048;
+
+  private static Schema getSchema() {
+    return new Schema(Arrays.asList(
+        Field.nullable("0", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("1", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("2", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("3", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("4", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("5", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("6", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("7", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("8", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("9", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)),
+        Field.nullable("10", new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE))
+    ));
+  }
+
+  /**
+   * Ensure that if the client cancels, the server does not leak memory.
+   *
+   * <p>In gRPC, canceling the stream from the client sends an event to the server. Once processed, gRPC will start
+   * silently rejecting messages sent by the server. However, Flight depends on gRPC processing these messages in order
+   * to free the associated memory.
+   */
+  @Test
+  public void testCancelingDoGetDoesNotLeak() {
+    final AtomicBoolean callFinished = new AtomicBoolean();
+    try (final BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+        final FlightServer s =
+            FlightTestUtil.getStartedServer(
+                (location) -> FlightServer.builder(allocator, location, new LeakFlightProducer(allocator, callFinished))
+                    .build());
+        final FlightClient client = FlightClient.builder(allocator, s.getLocation()).build()) {
+
+      final FlightStream stream = client.getStream(new Ticket(new byte[0]));
+      stream.getRoot();
+      stream.cancel("Cancel", null);
+
+      // Wait for the call to finish. (Closing the allocator while a call is ongoing is a guaranteed leak.)
+      while (!callFinished.get()) {
+        Thread.sleep(100);
+      }
+
+      s.shutdown();
+      s.awaitTermination();
+    } catch (Exception e) {
```

Comment:
I think its generally more typical to declare the test method as throwing Exception (or the specific ones that are thrown) then wrapping it here, is there a specific reason you are rethrowing?

Reply:
No reason, I've changed it.